File: algorithms/[503]下一个更大元素 II/solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
const nextGreaterElements = function(nums) {
  const n = nums.length;
  const stack = [];
  const res = [];

  for (let i = 2 * n - 1; i >= 0; i -= 1) {
    while (stack.length && nums[i % n] >= stack[stack.length - 1]) {
      stack.pop();
    }
    res[i % n] = stack.length ? stack[stack.length - 1] : -1;
    stack.push(nums[i % n]);
  }
  return  res;
};

File: algorithms/[503]下一个更大元素 II/solution.py
# topics = ["栈"]

from typing import List


class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res: List[int] = [-1] * n
        stack: List[int] = []

        for i in range(n * 2 - 1):
            while stack and nums[stack[-1]] < nums[i % n]:
                res[stack.pop()] = nums[i % n]
            stack.append(i % n)

        return res
